source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1428 @@
+  bool use_delimiter_ = false;
+};
+
+// Creates spv_text object containing text from |str|.
+// The returned value is owned by the caller and needs to be destroyed with
+// spvTextDestroy.
+spv_text CreateSpvText(const std::string& str) {
+  spv_text out = new spv_text_t();
+  assert(out);
+  char* cstr = new char[str.length() + 1];
+  assert(cstr);
+  std::strncpy(cstr, str.c_str(), str.length());
+  cstr[str.length()] = '\0';
+  out->str = cstr;
+  out->length = str.length();
+  return out;
+}
+
+// Base class for MARK-V encoder and decoder. Contains common functionality
+// such as:
+// - Validator connection and validation state.
+// - SPIR-V grammar and helper functions.
+class MarkvCodecBase {
+ public:
+  virtual ~MarkvCodecBase() {
+    spvValidatorOptionsDestroy(validator_options_);
+  }
+
+  MarkvCodecBase() = delete;
+
+  void SetModel(const MarkvModel* model) {
+    model_ = model;
+  }
+
+ protected:
+  struct MarkvHeader {
+    MarkvHeader() {
+      magic_number = kMarkvMagicNumber;
+      markv_version = GetMarkvVersion();
+      markv_model = 0;
+      markv_length_in_bits = 0;
+      spirv_version = 0;
+      spirv_generator = 0;
+    }
+
+    uint32_t magic_number;
+    uint32_t markv_version;
+    // Magic number to identify or verify MarkvModel used for encoding.
+    uint32_t markv_model;
+    uint32_t markv_length_in_bits;
+    uint32_t spirv_version;
+    uint32_t spirv_generator;
+  };
+
+  explicit MarkvCodecBase(spv_const_context context,
+                          spv_validator_options validator_options)
+      : validator_options_(validator_options),
+        vstate_(context, validator_options_), grammar_(context),
+        model_(GetDefaultModel()) {}
+
+  // Validates a single instruction and updates validation state of the module.
+  spv_result_t UpdateValidationState(const spv_parsed_instruction_t& inst) {
+    return ValidateInstructionAndUpdateValidationState(&vstate_, &inst);
+  }
+
+  // Returns the current instruction (the one last processed by the validator).
+  const Instruction& GetCurrentInstruction() const {
+    return vstate_.ordered_instructions().back();
+  }
+
+  // Returns true if the opcode has a variable number of operands.
+  bool OpcodeHasVariableNumberOfOperands(SpvOp opcode) const {
+    spv_opcode_desc opcode_desc;
+    if (grammar_.lookupOpcode(static_cast<SpvOp>(opcode), &opcode_desc)
+        != SPV_SUCCESS) {
+      assert(false && "Opcode not described in the grammar");
+    }
+
+    if (opcode_desc->numTypes == 0)
+      return false;
+
+    return spvOperandIsVariable(
+        opcode_desc->operandTypes[opcode_desc->numTypes - 1]);
+  }
+
+  spv_validator_options validator_options_;
+  ValidationState_t vstate_;
+  const libspirv::AssemblyGrammar grammar_;
+  MarkvHeader header_;
+  const MarkvModel* model_;
+
+  // Move-to-front list of all ids.
+  // TODO(atgoo@github.com) Consider a better move-to-front implementation.
```

Comment:
I've been designing a solution which would use timestamps and a rank/select tree. This way index-to-id becomes select element with k-th timestamp. Id-to-index would also need a hash map id-to-node, then rank(node) would be index.